### PR TITLE
Visual improvements in sublayers dialog

### LINKED ISSUE #26905: Display file names when loading multiple GPX files to project
Author Name: **Carolyn Krause** (Carolyn Krause)
Original Redmine Issue: [19075](https://issues.qgis.org/issues/19075)

Redmine category:gui

---
When adding a GPX file to a project, a popup window asks the user to select which types of layers to add from the GPX file. When adding multiple GPX files at a time, this dialog popups up once for each file, but there's no way to tell which file the current popup is referring to. 

Is it possible to add the name of the current GPX file to the top of the popup? Like in the attached image.

---

- [add gpx layers popup.png](https://issues.qgis.org/attachments/download/12761/add%20gpx%20layers%20popup.png) (Carolyn Krause)

### PR DESCRIPTION
**Pull request by suricactus**
An annoying lack of context when adding multiple layers:
1) select multiple multi-layer geopackage layers from a directory in the file browser
2) drag and drop them in QGIS
3) the "Select Layers to Add..." dialog does not give any hint about the source file

This PR fixes 3) by adding the current file name in the window title.
Peek 2020-03-14 01-11

Fixes: https://github.com/qgis/QGIS/issues/26905

### CODE REVIEW

**Comment by elpaso** [comment]
Nice! 
What do you think about hiding the path and showing the filename only? 
I'm worried about long paths.

**Comment by suricactus** [comment]
Imagine a situation you import files from a file tree, where you have multiple files with the same name. 
```
./census/2001/filename.gpkg
./census/2011/filename.gpkg
./census/2021/filename.gpkg
```
We should be covered  for this situation, I don't think it's that uncommon.

A possible solution would be to use, so we can shorten the path in most cases:
```
QDir( QgsProject::instance().homePath() ).relativeFilePath("D:/images/image.bmp");
```
Unfortunately storing a list of the current filenames to be shown is not that easy and trivial. I will add something like this is in `handleDropUriList`, where the suggested filename is generated.

Any other solutions is appreciated.

**Comment by suricactus** [comment]
A second though of the use case described in my post above. In such situation the files are most likely to have the same structure and one would like to import the same layers. This is just an excuse, the truth is I don't find an easy way to fix. The idea to use `relativeFilePath` will produce ugly "../../../../" strings and you are right that `/home/suricactus/projects/very long project name/layer/census/2010 unofficial/demographics/by age and gender/united states of america.gpkg` will look terrible in the title bar. 

I am using this branch to apply other small fixes. Summary:
- Fixes https://github.com/qgis/QGIS/issues/26905 **Show currently added filename in the titlebar**.
- Fixes https://github.com/qgis/QGIS/issues/26670 **Remove the filename prefix in the "Layer Tree"**. Useless visual clutter. Putting all the layer in a group is already present, so there is no risk of mixing layer names together. Adding this as an additional checkbox in the dialog causes will make the interface to look complicated.
- **Disable the OK button if there are no layers selected.** Most users would just click "OK" and would be confused why there is no effect.
- **Select all layers by default.** To finish the operation is exactly one mouse click of the "OK" button, as it is a common case to "Select All". Otherwise the user is always obliged to click on a layer or "Select All" and then to click "OK". 

Breaking changes:
- setting `qgis/promptForRasterSublayers` is renamed to `qgis/promptForSublayers`. The default value is still "Always". I renamed the setting to something sensible, there will be minor regression for some users who changed this value. The setting is renamed to "Prompt for ~raster~ sublayers when opening". It seems that this option either never existed in the UI, or has disappeared somehow.

Visual changelog:
- Settings:

image

- In action:
Peek 2020-03-17 23-14

Known issues:
- **Add option to "Add to a group" zipfiles** - this option is a bit tricky to add, I have overload several functions. I would prefer to do that in a separate PR.

**Comment by uclaros** [comment]
Regarding the long paths, a read only widget could be added above the table view (a grayed out lineedit?) like and address bar so you can also select and copy part of the path-filename.

**Comment by suricactus** [comment]
I don't see the point of being able to copy your file path in 99% of the cases? You only need indication what are you currently seeing, no need to make the interface heavier than needed with add additional boxes in my opinion.

**Comment by uclaros** [comment]
- It's not about copying it, it's about seeing it as you have already mentioned. Imho, having this info only on the titlebar is not a good choice, since the titlebar is customizable by the window manager it is not guaranteed that what is there is readable. Adding a single line of text is not going to make this simple dialog bloated. And in general it is good practice to be able to copy displayed paths. If it's there, why not allow copying it for the 1% case?

- If `PromptNever` is chosen it silently skips loading a file? Do we need that?

**Comment by suricactus** [comment]
1) fair point on window managers. I will change that.
2) this was already existing behavior. I see it as a way to disable drag and drop of layer files in QGIS.

**Comment by suricactus** [comment]
Looks a bit ugly, but now the user knows exactly what is being imported. 
Peek 2020-03-19 06-31

**Comment by uclaros** [comment]
Removing the 'Path' QLabel and letting the QLineEdit occupy the full width will definitely make it look better.
You could also leave the filename in the titlebar too.

**Comment by suricactus** [comment]
This will need changes here:
https://docs.qgis.org/testing/en/docs/user_manual/introduction/general_tools.html#identify

**Comment by uclaros** [comment]
> This will need changes here:
> https://docs.qgis.org/testing/en/docs/user_manual/introduction/general_tools.html#identify

I guess this belongs to PR #35190

**Comment by suricactus** [comment]
It took me some time to find the documentation error, the message is a bit hidden, even though colored in yellow.

@uclaros thanks for correction, something happened and I had a lot of comments posted in the wrong threads.

As for this PR:
- I agree it looks better without the "Path" label, so I removed it
- Put filename only in the window title

Yet another iteration of this gif:
Peek 2020-03-25 07-36

**Comment by uclaros** [comment]
It looks very good! Is text selectable in the qlineedit?

**Comment by suricactus** [comment]
Now it is. :)

**Review by m-kuhn — COMMENTED** [review]
If others have already left their :heart: for the UX it provides. Here are some code (style) level things to check and then this should be good to roll :+1:

File: src/gui/qgssublayersdialog.cpp [hunk 1]
```diff
@@ -61,27 +63,34 @@ QgsSublayersDialog::QgsSublayersDialog( ProviderType providerType, const QString
...
-    setWindowTitle( tr( "Select Layers to Add…" ) );
+    title = tr( "Select Layers to Add…" );
     layersTable->setHeaderLabels( QStringList() << tr( "Layer ID" ) << tr( "Layer name" )
                                   << tr( "Type" ) );
     mShowType = true;
   }
 
+  QString fileFullPath = providerType == QgsSublayersDialog::Vsifile
+                         ? providerSource
+                         : QgsProviderRegistry::instance()->decodeUri( name, providerSource )[QStringLiteral( "path" )].toString();
```
[the remark this answers is not in the document]
Comment (suricactus) [reply src/gui/qgssublayersdialog.cpp:79]: done!

File: src/gui/qgssublayersdialog.h [hunk 2]
```diff
@@ -93,7 +122,7 @@ class GUI_EXPORT QgsSublayersDialog : public QDialog, private Ui::QgsSublayersDi
      * If we should add layers in a group
      * \since QGIS 3.0
      */
-    bool addToGroupCheckbox() const { return mCheckboxAddToGroup->isChecked(); }
+    bool addToGroupCheckbox() const { return cbxAddToGroup->isChecked(); }
```
[the remark this answers is not in the document]
Comment (suricactus) [reply src/gui/qgssublayersdialog.h:125]: This is quite inconsistent across the codebase though. Renamed it to `mCbxAddToGroup`

**Comment by suricactus** [comment]
@m-kuhn is it possible to reopen the PR or I have to make a new one?

File: src/gui/qgssublayersdialog.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (m-kuhn) [reply src/gui/qgssublayersdialog.h:125]: "organically grown" is how you could call it...

[the remark this answers is not in the document]
Comment (suricactus) [reply src/gui/qgssublayersdialog.h:125]: How far should I go for this? Is this convention valid only for elements that are used in the code? There are quite a lot of GUI elements, that we never interact in "hand-written" code like layouts, separators, dialog buttons etc. Should we also change them?

I believe this organically grown change would be more widely applied if it is described somewhere. If you can point me documentation/style guide/github page/chat where this has been decided and described, would be nice. I would love to write these two sentences describing the requirement, but it is still unclear how it should be applied. Maybe here is a good candidate to put such recommendation? https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#qt-designer

[the remark this answers is not in the document]
Comment (m-kuhn) [reply src/gui/qgssublayersdialog.h:125]: > Is this convention valid only for elements that are used in the code? There are quite a lot of GUI elements, that we never manually use in the code like layouts, separators, dialog buttons etc. Should we also change them?

Yes, only for the ones used in code

> Maybe here is a good candidate to put such recommendation? https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#qt-designer

That's exactly where it should go, would be really nice to add that there!

**Pull request closed without merging**